prow/gerrit/adapter/trigger.go

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package adapter
+
+import (
+	"k8s.io/test-infra/prow/pjutil"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/gerrit/client"
+)
+
+const layout = "2006-01-02 15:04:05"
+
+// messageFilter builds a filter for jobs based on the messageBody matching the trigger regex of the jobs.
+func messageFilter(lastUpdate time.Time, change client.ChangeInfo, presubmits []config.Presubmit) (pjutil.Filter, error) {
+	var filters []pjutil.Filter
+	currentRevision := change.Revisions[change.CurrentRevision].Number
+	for _, message := range change.Messages {
+		messageTime, err := time.Parse(layout, message.Date)
+		if err != nil {
+			logrus.WithError(err).Errorf("Parse time %v failed", message.Date)
+			continue
+		}
+		if message.RevisionNumber != currentRevision || messageTime.Before(lastUpdate) {
+			continue
+
+		}
+		// Skip comments not germane to this plugin
+		if !pjutil.TestAllRe.MatchString(message.Message) {
+			matched := false
+			for _, presubmit := range presubmits {
+				matched = matched || presubmit.TriggerMatches(message.Message)
+				if matched {
+					filters = append(filters, pjutil.CommandFilter(message.Message))
+				}
+			}
+			if !matched {
+				logrus.Infof("Comment %s doesn't match any triggering regex, skipping.", message.Message)
+				continue
+			}
+
+		} else {
+			filters = append(filters, pjutil.TestAllFilter())
+		}
+
+	}
+	return pjutil.AggregateFilter(filters), nil
+}
+
+// oldRevisionFilter builds a filter for jobs that have already been processed for a revision.
+func oldRevisionFilter(lastUpdate time.Time, rev client.RevisionInfo) pjutil.Filter {
+	created, err := time.Parse(layout, rev.Created)
+	if err != nil {
+		logrus.WithError(err).Errorf("Parse time %v failed", rev.Created)
+		return func(p config.Presubmit) (bool, bool, bool) {
+			return false, false, false
+		}
+	}
+
+	return func(p config.Presubmit) (bool, bool, bool) {
+		return created.After(lastUpdate), false, false
```

Comment:
This is needed to filter out old changes that have new messages but don't match any trigger.

Reply:
This isn't filtering changes though, this is a ProwJob filter that doesn't even look at the ProwJob. It always returns true or always returns false for all inputs. Maybe I'm missing something, but I don't really see the point of a filter that always returns true (just don't run the filter) or always returns false (just return an empty slice).

Reply:
Yeah, I agree we don't necessarily need a filter here: I had something similar earlier https://github.com/kubernetes/test-infra/pull/11448#discussion_r260050976 . Should I get rid of this filter and just add that check back?

Reply:
I can't really see what the code was like before because it is outdated now so I'm not sure what Steve's concern was. I suspect it was related to the lack of filters or apparent special casing?

IMO this logic would make way more sense as a check than a filter since we aren't actually filtering at all, but I also think I am either still misunderstanding or the logic is wrong.

If the revision is new since the last update time then we should use a `TestAllFilter` because all the always_run jobs need to be triggered. If the revision is not newer than the last update time then we should not add a filter at all. So I'd expect this to just be a check that optionally adds an extra filter to the slice of filters.

Reply:
done